### PR TITLE
HIVE-25975: Optimize ClusteredWriter for bucketed Iceberg tables

### PR DESCRIPTION
**Pull request by szlta**
This adds a new UDF that uses Iceberg's bucket transformation function to produce bucket values from constants or any column input. All types that Iceberg buckets support are supported in this UDF too, except for UUID.

This UDF is then used in SortedDynPartitionOptimizer to sort data during write if the target Iceberg target has bucket transform partitioning. 

To enable this, Hive has been extended with the feature that allows storage handlers to define custom sorting expressions, to be passed to FileSink operator's DynPartContext during dynamic partitioning write scenarios.

The lenient version of ClusteredWriter in patched-iceberg-core has been disposed of as it is not needed anymore with this feature in.

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+import org.apache.iceberg.types.Types;
+
+/**
+ * GenericUDFIcebergBucket - UDF that wraps around Iceberg's bucket transform function
+ */
+@Description(name = "iceberg_bucket",
+    value = "_FUNC_(value, bucketCount) - " +
+        "Returns the bucket value calculated by Iceberg bucket transform function ",
+    extended = "Example:\n  > SELECT _FUNC_('A bucket full of ice!', 5);\n  4")
+//@VectorizedExpressions({StringLength.class})
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:53]: nit: Why is this commented out? Shall we just remove it?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            " ICEBERG_BUCKET() only takes STRING/CHAR/VARCHAR/BINARY/INT/LONG/DECIMAL/FLOAT/DOUBLE" +
+                " types as first argument, got " + inputType);
+    }
+
+    argumentOI = (PrimitiveObjectInspector) arguments[1];
+    if (!isValid2ndArgumentType(arguments[1])) {
+      throw new UDFArgumentTypeException(1,
+          "ICEBERG_BUCKET() second argument only takes an int types, got " + arguments[1].getTypeName());
+    }
+    numBuckets = ((WritableConstantIntObjectInspector) argumentOI).getWritableConstantValue().get();
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:140]: Question: It might be just me, but instead of `isValid2ndArgumentType` I would create a `private int getNumBuckets(ObjectInspector) throws UDFArgumentException` method

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+  @Override
+  public Object evaluate(DeferredObject[] arguments) throws HiveException {
+
+    DeferredObject argument = arguments[0];
+    if (argument == null) {
+      return null;
+    }
+
+    if (Types.StringType.get().equals(icebergType)) {
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:166]: Can we do this with a method, so we do not have to run the whole if/else misey?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 4]
```diff
@@ -96,6 +104,20 @@
   private static final String ICEBERG_URI_PREFIX = "iceberg://";
   private static final Splitter TABLE_NAME_SPLITTER = Splitter.on("..");
   private static final String TABLE_NAME_SEPARATOR = "..";
+  private static final transient BiFunction<Integer, Integer, Function<List<ExprNodeDesc>, ExprNodeDesc>>
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:107]: nit: Comment please

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 5]
```diff
@@ -307,6 +328,41 @@ public boolean supportsPartitionTransform() {
...
 
+  @Override
+  public DynamicPartitionCtx createDPContext(HiveConf hiveConf, org.apache.hadoop.hive.ql.metadata.Table hmsTable)
+      throws SemanticException {
+    TableDesc tableDesc = Utilities.getTableDesc(hmsTable);
+    Table table = IcebergTableUtil.getTable(conf, tableDesc.getProperties());
+    if (table.spec().isUnpartitioned()) {
+      return null;
+    }
+    // Iceberg currently doesn't have publicly accessible partition transform information, hence use above string parse
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:339]: nit: newline

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java [hunk 6]
```diff
@@ -280,6 +283,21 @@ default boolean supportsPartitionTransform() {
     return null;
   }
 
+  /**
+   * Creates a DynnamicPartitionCtx instance that will be set up by the storage handler itself. Useful for non-native
+   * tables where partitions are not handled by Hive, and sorting is required in a custom way before writing the table.
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java:288]: What happens if the return value is `null`

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 7]
```diff
@@ -307,6 +328,41 @@ public boolean supportsPartitionTransform() {
...
+    List<PartitionTransformSpec> partitionTransformSpecs = getPartitionTransformSpec(hmsTable);
+
+    DynamicPartitionCtx dpCtx = new DynamicPartitionCtx(new LinkedHashMap<>(),
+        hiveConf.getVar(HiveConf.ConfVars.DEFAULTPARTITIONNAME),
+        hiveConf.getIntVar(HiveConf.ConfVars.DYNAMICPARTITIONMAXPARTSPERNODE));
+    List<Function<List<ExprNodeDesc>, ExprNodeDesc>> customSortExprs = new LinkedList<>();
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:345]: Nit: Lists.newLinkedList()

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 8]
```diff
@@ -307,6 +328,41 @@ public boolean supportsPartitionTransform() {
...
+    dpCtx.setCustomSortExpressions(customSortExprs);
+
+    Map<String, Integer> fieldOrderMap = new HashMap<>();
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:348]: Nit: Maps.newHashMap()

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 9]
```diff
@@ -307,6 +328,41 @@ public boolean supportsPartitionTransform() {
...
+    List<Types.NestedField> fields = table.schema().columns();
+    for (int i = 0; i < fields.size(); ++i) {
+      fieldOrderMap.put(fields.get(i).name(), i);
+    }
+
+    for (PartitionTransformSpec spec : partitionTransformSpecs) {
+      int order = fieldOrderMap.get(spec.getColumnName());
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:355]: Is it possible to create 2 transforms for a single column?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/FunctionRegistry.java [hunk 10]
```diff
@@ -585,6 +585,13 @@
     system.registerGenericUDF(GenericUDFMaskShowFirstN.UDF_NAME, GenericUDFMaskShowFirstN.class);
     system.registerGenericUDF(GenericUDFMaskShowLastN.UDF_NAME, GenericUDFMaskShowLastN.class);
     system.registerGenericUDF(GenericUDFMaskHash.UDF_NAME, GenericUDFMaskHash.class);
+
+    try {
+      system.registerGenericUDF("iceberg_bucket",
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/FunctionRegistry.java:590]: What happens if we add the iceberg jar with the `ADD JAR` command?

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java [hunk 11]
```diff
@@ -280,6 +283,21 @@ default boolean supportsPartitionTransform() {
     return null;
   }
 
+  /**
+   * Creates a DynnamicPartitionCtx instance that will be set up by the storage handler itself. Useful for non-native
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java:287]: nit: `DynnamicPartitionCtx`-> `DynamicPartitionCtx`

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionOptimizer.java [hunk 12]
```diff
@@ -169,17 +182,27 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
...
+      if (allStaticPartitions(fsParent, allRSCols, dpCtx)) {
         LOG.debug("Bailing out of sorted dynamic partition optimizer as all dynamic partition" +
             " columns got constant folded (static partitioning)");
         return null;
       }
 
-      DynamicPartitionCtx dpCtx = fsOp.getConf().getDynPartCtx();
       List<Integer> partitionPositions = getPartitionPositions(dpCtx, fsParent.getSchema());
+      LinkedList<Function<List<ExprNodeDesc>, ExprNodeDesc>> customSortExprs =
+          new LinkedList<>(dpCtx.getCustomSortExpressions());
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionOptimizer.java:202]: Nit: Lists.newLinkedList?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionOptimizer.java [hunk 13]
```diff
@@ -521,34 +570,45 @@ private void inferSortPositions(Operator<? extends OperatorDesc> fsParent,
...
-        Operator<? extends OperatorDesc> parent, AcidUtils.Operation writeType) throws SemanticException {
+    public ReduceSinkOperator getReduceSinkOp(List<Integer> partitionPositions, List<Integer> sortPositions,
+        List<Function<List<ExprNodeDesc>, ExprNodeDesc>> customSortExprs, List<Integer> sortOrder,
+        List<Integer> sortNullOrder, ArrayList<ExprNodeDesc> allCols, ArrayList<ExprNodeDesc> bucketColumns,
+        int numBuckets, Operator<? extends OperatorDesc> parent, AcidUtils.Operation writeType) {
+
+      // Order of KEY columns, if custom sort is present partition and bucket columns are disregarded:
+      // 0) Custom sort expressions
+      //                              1) Partition columns
+      //                              2) Bucket number column
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionOptimizer.java:581]: nit: formatting

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionOptimizer.java [hunk 14]
```diff
@@ -521,34 +570,45 @@ private void inferSortPositions(Operator<? extends OperatorDesc> fsParent,
...
+        numBuckets = -1;
+      }
+
       keyColsPosInVal.addAll(partitionPositions);
       if (bucketColumns != null && !bucketColumns.isEmpty()) {
         keyColsPosInVal.add(-1);
       }
       keyColsPosInVal.addAll(sortPositions);
 
+
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionOptimizer.java:603]: nit: not needed

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+  private transient PrimitiveObjectInspectorConverter.FloatConverter floatConverter;
+  private transient PrimitiveObjectInspectorConverter.DoubleConverter doubleConverter;
+  private transient Type.PrimitiveType icebergType;
+  private int numBuckets = -1;
+
+  @Override
+  public ObjectInspector initialize(ObjectInspector[] arguments) throws UDFArgumentException {
+    if (arguments.length != 2) {
+      throw new UDFArgumentLengthException(
+          "ICEBERG_BUCKET requires 2 argument, got " + arguments.length);
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:71]: nit: arguments
also, maybe we should tell the user what those arguments are, i.e. (value, bucketCount)?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+          "ICEBERG_BUCKET() second argument only takes an int types, got " + arguments[1].getTypeName());
+    }
+    numBuckets = ((WritableConstantIntObjectInspector) argumentOI).getWritableConstantValue().get();
+
+    outputOI = PrimitiveObjectInspectorFactory.writableIntObjectInspector;
+    return outputOI;
+  }
+
+  private boolean isValid2ndArgumentType(ObjectInspector arg) {
+    if (arg.getCategory() != ObjectInspector.Category.PRIMITIVE) {
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:147]: nit: maybe just
```
return arg.getCategory() == ObjectInspector.Category.PRIMITIVE && 
    argumentOI.getPrimitiveCategory() == PrimitiveObjectInspector.PrimitiveCategory.INT
```

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+      throw new UDFArgumentTypeException(1,
+          "ICEBERG_BUCKET() second argument only takes an int types, got " + arguments[1].getTypeName());
+    }
+    numBuckets = ((WritableConstantIntObjectInspector) argumentOI).getWritableConstantValue().get();
+
+    outputOI = PrimitiveObjectInspectorFactory.writableIntObjectInspector;
+    return outputOI;
+  }
+
+  private boolean isValid2ndArgumentType(ObjectInspector arg) {
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:146]: Would it makes sense to do all argument validation in one piece at the beginning? e.g. `validateArgs(arguments)`

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 18]
```diff
@@ -307,6 +328,41 @@ public boolean supportsPartitionTransform() {
...
+      throws SemanticException {
+    TableDesc tableDesc = Utilities.getTableDesc(hmsTable);
+    Table table = IcebergTableUtil.getTable(conf, tableDesc.getProperties());
+    if (table.spec().isUnpartitioned()) {
+      return null;
+    }
+    // Iceberg currently doesn't have publicly accessible partition transform information, hence use above string parse
+    List<PartitionTransformSpec> partitionTransformSpecs = getPartitionTransformSpec(hmsTable);
+
+    DynamicPartitionCtx dpCtx = new DynamicPartitionCtx(new LinkedHashMap<>(),
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:342]: nit: Maps.newLinkedHashMap()

File: ql/src/java/org/apache/hadoop/hive/ql/plan/DynamicPartitionCtx.java [hunk 19]
```diff
@@ -50,6 +52,14 @@
...
   private int maxPartsPerNode;    // maximum dynamic partitions created per mapper/reducer
   private Pattern whiteListPattern;
+  /**
+   * Expressions describing a custom way of sorting the table before write. Expressions can reference simple
+   * column descriptions or a tree of expressions containing more columns and UDFs.
+   * Can be useful for custom bucket/hash sorting.
+   * A custom expression should be a lambda that is given the original column description expressions as per read
+   * schema and returns a single expression. Example for simply just referencing column 3: cols -> cols.get(3).clone()
+   */
+  private transient List<Function<List<ExprNodeDesc>, ExprNodeDesc>> customSortExpressions;
```
Comment (marton-bod) [inline ql/src/java/org/apache/hadoop/hive/ql/plan/DynamicPartitionCtx.java:62]: Out of curiosity, why transient?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 20]
```diff
@@ -8128,13 +8128,9 @@ private FileSinkDesc createFileSinkDesc(String dest, TableDesc table_desc,
...
       if (storageHandler != null && storageHandler.alwaysUnpartitioned()) {
-        List<PartitionTransformSpec> nonNativePartSpecs = storageHandler.getPartitionTransformSpec(dest_tab);
-        if (dpCtx == null && nonNativePartSpecs != null && !nonNativePartSpecs.isEmpty()) {
-          verifyDynamicPartitionEnabled(conf, qb, dest);
-          Map<String, String> partSpec = new LinkedHashMap<>();
-          nonNativePartSpecs.forEach(ps -> partSpec.put(ps.getColumnName(), null));
-          dpCtx = new DynamicPartitionCtx(partSpec, conf.getVar(HiveConf.ConfVars.DEFAULTPARTITIONNAME),
-              conf.getIntVar(HiveConf.ConfVars.DYNAMICPARTITIONMAXPARTSPERNODE));
+        DynamicPartitionCtx nonNativeDpCtx = storageHandler.createDPContext(conf, dest_tab);
+        if (dpCtx == null && nonNativeDpCtx != null) {
```
Comment (marton-bod) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:8132]: This means `dpCtx` does not get initialized at this point for Iceberg queries?

File: ql/src/java/org/apache/hadoop/hive/ql/plan/DynamicPartitionCtx.java (hunk 19 above)
Comment (szlta) [reply ql/src/java/org/apache/hadoop/hive/ql/plan/DynamicPartitionCtx.java:62]: The job has the dpCtx serialized, and for some reason kryo is throwing exceptions for lambdas on the executor side.
The expressions here are only needed in HS2 during query planning so I just prevent them from going to executors in the first place.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java (hunk 1 above)
Comment (szlta) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:53]: just a leftover - removing it

File: ql/src/java/org/apache/hadoop/hive/ql/plan/DynamicPartitionCtx.java (hunk 19 above)
Comment (marton-bod) [reply ql/src/java/org/apache/hadoop/hive/ql/plan/DynamicPartitionCtx.java:62]: Yeah, I was wondering if it's okay if these expressions don't get to the executors. But I got my answer - Thanks!

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java (hunk 3 above)
Comment (szlta) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:166]: Not sure how you mean - originally I wanted a switch, but the Iceberg types are unavailable as enums. But then again for some types the logic is different, like bytes, decimals.

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java (hunk 6 above)
Comment (szlta) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java:288]: fixing javadoc

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 9 above)
Comment (szlta) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:355]: Sure, why not? Those expressions would just have the same source column.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 20 above)
Comment (szlta) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:8132]: That's right. It's because Hive doesn't know about partition column in Iceberg tables, so doesn't bother creating a dynamic partitioning context at all..

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java (hunk 3 above)
Comment (szlta) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:165]: Discussed offline - will be refactoring

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java (hunk 2 above)
Comment (szlta) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:139]: makes sense

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java (hunk 17 above)
Comment (szlta) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:145]: This will be refactored into getNumBuckets

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+      return null;
+    } else {
+      evaluator.apply(argument);
+    }
+
+    return result;
+  }
+
+  private <T> void applyBucketTransform(T value, Type.PrimitiveType icebergType) {
+    Transform<T, Integer> transform = Transforms.bucket(icebergType, numBuckets);
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java:190]: Would it make sense to create the transform in the `initiaiize` method instead of recreating it for every row?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergOutputCommitter.java [hunk 22]
```diff
@@ -77,7 +77,7 @@
   );
 
   private static final PartitionSpec PARTITIONED_SPEC =
-      PartitionSpec.builderFor(CUSTOMER_SCHEMA).bucket("customer_id", 3).build();
+      PartitionSpec.builderFor(CUSTOMER_SCHEMA).identity("customer_id").build();
```
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergOutputCommitter.java:80]: I remember that you have already told me why is this change needed. Could you please refresh my memory?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java (hunk 21 above)
Comment (szlta) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java]: Yes it would, I just refactored this. I must say I really liked my generic method though :/

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergOutputCommitter.java (hunk 22 above)
Comment (szlta) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergOutputCommitter.java:80]: Most of the write path is mocked out, Hive queries are not committed to test writes in this test class. That means no optimizers can run either, and thus the result set before a write would not be sorted by bucket function. With the absence of lenient ClusteredWriter this would make the test always fail for partitioned writes.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 23]
```diff
@@ -307,6 +331,42 @@ public boolean supportsPartitionTransform() {
...
+  public DynamicPartitionCtx createDPContext(HiveConf hiveConf, org.apache.hadoop.hive.ql.metadata.Table hmsTable)
+      throws SemanticException {
+    TableDesc tableDesc = Utilities.getTableDesc(hmsTable);
+    Table table = IcebergTableUtil.getTable(conf, tableDesc.getProperties());
+    if (table.spec().isUnpartitioned()) {
+      return null;
+    }
+
+    // Iceberg currently doesn't have publicly accessible partition transform information, hence use above string parse
+    List<PartitionTransformSpec> partitionTransformSpecs = getPartitionTransformSpec(hmsTable);
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:344]: You might want beat me with a stick for mentioning this only now, but `getPartitionTransformSpec(hmsTable)` starts with:
```
    TableDesc tableDesc = Utilities.getTableDesc(hmsTable);
    Table table = IcebergTableUtil.getTable(conf, tableDesc.getProperties());
```
We have the Iceberg table at hand. Would it be better to have a `private partitionTransformSpec(Table iceTable)` which could be called here, and from `getPartitionTransformSpec`?

**Review by pvary — APPROVED** [review]
One minimal comment - if you have to rerun the tests again, please check it out, but not really needed for merge.

+1

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 23 above)
Comment (szlta) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:344]: As discussed offline this won't be addressed with this patch.
